### PR TITLE
Feat/add admin otp flow

### PR DESCRIPTION
**Pull request by kwajiehao**
## Overview

Currently, the OTP flow is catered to users under quarantine - their data can be found in the `QuarantineOrders` table. However, we need to use the OTP flow for admin users as well since they need to access our APIs for their dashboards. This PR helps to set up an OTP flow that accommodates OTP for both quarantine ordered users as well as admin users.

The main changes are:
- Creation of `AdminUsers` model and accompanying migration
- Updating `OtpService` methods to accomodate both phone numbers from quarantine ordered-people and emails from admin users
- Update OTP model to accommodate `email`, make `contact_number` field nullable, plus accompanying migration
- Add seed data for `AdminUsers`
- Modify auth middleware to distinguish between admin users'  access tokens and quarantine ordered-users' access tokens

### Add an `AdminUsers` model and `AdminUserService`
An `AdminUsers` model was created with an `email` column for which all values have to be unique. The accompanying migration was created using the sequelize CLI's `model:create` command.

The `AdminUserService` was created so that we can check by their email whether an admin user exists or not. This is used in the validation step for OTP generation.

### Update OTP model to accommodate `email`
The OTP model was updated to accommodate an `email` column.  Since an OTP can be associated with either a phone number or an email, the `email` column is allowed to be null and the `contact_number` column was changed so that it is now allowed to be null. Finally, an additional migration was applied  to add the `email` column and change the `contact_number` column.

### Update OTP route functions
Both the `generateOtp` and `verifyOtp` route functions were modified to accommodate both emails and contact numbers. This is done by validating the `contact` param and checking whether it is a phone number or a `gov.sg` email. If it's a phone number, it's validated with the `QuarantineOrderService` and if it's an email it is validated with the `AdminUserService`. 

The `verifyOtp` route function is also updated to return an access token with a `role` attribute with a value of `admin` instead of the `orderId` attribute which is returned for quarantined users.

### Add seed data for the `AdminUsers` table
This is self-explanatory

### Modify auth middleware to verify admin users
The auth middleware now accepts a param, `isAdmin`, which has a default value of `false`. If `isAdmin` is true, it returns a middleware function which checks the access token for a `role` attribute with the value `admin`. Otherwise, it simply checks whether the JWT is valid.

### CODE REVIEW

File: app/routes/otp.js [hunk 1]
```diff
@@ -26,20 +51,39 @@ const { TOKEN_SIGNING_KEY } = process.env;
...
     }
     while (otp.match(/[a-z]/i));
-    console.log(`Generated OTP ${otp} for ${contactNumber}`);
+    console.log(`Generated OTP ${otp} for ${contact}`);
 
     const hashedOtp = bcrypt.hashSync(otp, SALT_ROUNDS);
-    await OtpService.saveOtp(contactNumber, hashedOtp);
+
+    // 3. Save OTP to table
+    await OtpService.saveOtp({
```
Comment (christabelneo) [inline app/routes/otp.js:81]:
```
let newOTP 
if (isAdmin) {
 newOTP = {
  null,
  contact,
  hashedOTP
 }
} else {
newOTP = {
  contact,
  null,
  hashedOTP
 }
}

await OtpService.saveOtp(newOTP)
```

File: app/routes/otp.js [hunk 2]
```diff
@@ -7,15 +7,40 @@ const jwt = require('jsonwebtoken');
 const router = express.Router();
 const OtpService = require('../services/OtpService');
 const QuarantineOrderService = require('../services/QuarantineOrderService');
+const AdminUserService = require('../services/AdminUserService');
 
 const SALT_ROUNDS = 10;
 const { TOKEN_SIGNING_KEY } = process.env;
+const PHONE_NUMBER_REGEX = /^65[0-9]{8}$/;
+const GOV_SG_EMAIL_REGEX = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.+-]+\.gov.sg$/;
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply app/routes/otp.js]: Resolved with [8476835](https://github.com/datagovsg/homer-native-backend/pull/10/commits/8476835e26f976f919f2b2609a9b694ce21251e1)

File: app/routes/otp.js [hunk 3]
```diff
@@ -50,39 +94,62 @@ async function generateOtp(req, res) {
...
+
+    // 2. If quarantine order, retrieve latest order linked to phone number
+    // and assign orderId to the access token. If admin user, just assign the role
+    // admin.
+    let accessTokenParams;
+    if (isAdmin) {
+      const adminExpiry = new Date();
+      adminExpiry.setDate(adminExpiry.getDate() + ADMIN_USER_TOKEN_EXPIRY);
+      accessTokenParams = {
+        role: 'admin',
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply app/routes/otp.js:158]: Resolved with [8476835](https://github.com/datagovsg/homer-native-backend/pull/10/commits/8476835e26f976f919f2b2609a9b694ce21251e1)

File: app/routes/otp.js [hunk 4]
```diff
@@ -7,15 +7,40 @@ const jwt = require('jsonwebtoken');
...
 const QuarantineOrderService = require('../services/QuarantineOrderService');
+const AdminUserService = require('../services/AdminUserService');
 
 const SALT_ROUNDS = 10;
 const { TOKEN_SIGNING_KEY } = process.env;
+const PHONE_NUMBER_REGEX = /^65[0-9]{8}$/;
+const GOV_SG_EMAIL_REGEX = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.+-]+\.gov.sg$/;
+const ADMIN_USER_TOKEN_EXPIRY = 7; // 1 week in number of days
+
+const checkAdminUser = (contact) => {
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply app/routes/otp.js]: Resolved with [8476835](https://github.com/datagovsg/homer-native-backend/pull/10/commits/8476835e26f976f919f2b2609a9b694ce21251e1)

File: app/routes/otp.js [hunk 5]
```diff
@@ -26,20 +51,39 @@ const { TOKEN_SIGNING_KEY } = process.env;
  * No Quarantine Order found for this number
  */
 async function generateOtp(req, res) {
-  const { contact_number: contactNumber } = req.body;
+  const { contact } = req.body;
   try {
-    // 1. Check if user is on a quarantine order before issuing OTP
-    await QuarantineOrderService.getLatestOrderByContactNumber(contactNumber);
+    // 0. Validate whether contact is a phone number or email
+    const {
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply app/routes/otp.js]: Resolved with [8476835](https://github.com/datagovsg/homer-native-backend/pull/10/commits/8476835e26f976f919f2b2609a9b694ce21251e1)

File: app/routes/otp.js [hunk 6]
```diff
@@ -26,20 +51,39 @@ const { TOKEN_SIGNING_KEY } = process.env;
...
+      isAdmin,
+      email,
+      contactNumber,
+    } = checkAdminUser(contact);
+
+    // 1. Check if user is on a quarantine order or admin user before issuing OTP
+    if (isAdmin) {
```
[the remark this answers is not in the document]
Comment (kwajiehao) [reply app/routes/otp.js:76]: Resolved with [8476835](https://github.com/datagovsg/homer-native-backend/pull/10/commits/8476835e26f976f919f2b2609a9b694ce21251e1)

File: app/routes/otp.js (hunk 1 above)
Comment (kwajiehao) [reply app/routes/otp.js]: Resolved with [8476835](https://github.com/datagovsg/homer-native-backend/pull/10/commits/8476835e26f976f919f2b2609a9b694ce21251e1)

**Pull request merged**